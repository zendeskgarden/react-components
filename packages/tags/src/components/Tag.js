/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { StyledTag, StyledAvatar } from '../styled';
import Close from './Close';

const SIZE = {
  SMALL: 'small',
  MEDIUM: 'medium',
  LARGE: 'large'
};

const HUE = {
  GREY: 'grey',
  BLUE: 'blue',
  KALE: 'kale',
  RED: 'red',
  GREEN: 'green',
  YELLOW: 'yellow',
  FUSCHIA: 'fuschia',
  PINK: 'pink',
  CRIMSON: 'crimson',
  ORANGE: 'orange',
  LEMON: 'lemon',
  LIME: 'lime',
  MINT: 'mint',
  TEAL: 'teal',
  AZURE: 'azure',
  ROYAL: 'royal',
  PURPLE: 'purple'
};

/**
 * Accepts all `<div>` props.
 *
 * **Accessibility warning:** the `Tag.Close` component includes a default
 * English `aria-label`. If you are localizing your application you must
 * overwrite this property with your own translation.
 */
const Tag = React.forwardRef((props, ref) => <StyledTag ref={ref} {...props} />);

Tag.propTypes = {
  size: PropTypes.oneOf([SIZE.SMALL, SIZE.MEDIUM, SIZE.LARGE]),
  hue: PropTypes.oneOf([
    HUE.GREY,
    HUE.BLUE,
    HUE.KALE,
    HUE.RED,
    HUE.GREEN,
    HUE.YELLOW,
    HUE.FUSCHIA,
    HUE.PINK,
    HUE.CRIMSON,
    HUE.ORANGE,
    HUE.LEMON,
    HUE.LIME,
    HUE.MINT,
    HUE.TEAL,
    HUE.AZURE,
    HUE.ROYAL,
    HUE.PURPLE
  ]),
  pill: PropTypes.bool,
  round: PropTypes.bool,
  focused: PropTypes.bool
};

Tag.defaultProps = {
  size: SIZE.MEDIUM
};

Tag.Avatar = StyledAvatar;
Tag.Close = Close;

/** @component */
export default Tag;
