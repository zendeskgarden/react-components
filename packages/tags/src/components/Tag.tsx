/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLAttributes } from 'react';
import { StyledTag, StyledAvatar } from '../styled';
import Close from './Close';

interface ITagProps extends HTMLAttributes<HTMLDivElement> {
  size?: 'small' | 'medium' | 'large';
  hue?:
    | 'grey'
    | 'blue'
    | 'kale'
    | 'red'
    | 'green'
    | 'yellow'
    | 'fuschia'
    | 'pink'
    | 'crimson'
    | 'orange'
    | 'lemon'
    | 'lime'
    | 'mint'
    | 'teal'
    | 'azure'
    | 'royal'
    | 'purple';
  pill?: boolean;
  round?: boolean;
}

/**
 * Accepts all `<div>` props.
 *
 * **Accessibility warning:** the `Tag.Close` component includes a default
 * English `aria-label`. If you are localizing your application you must
 * overwrite this property with your own translation.
 */
const Tag = React.forwardRef<HTMLDivElement, ITagProps>(({ size, hue, ...otherProps }, ref) => (
  <StyledTag ref={ref} size={size} hue={hue} {...otherProps} />
));

Tag.defaultProps = {
  size: 'medium',
  hue: 'grey'
};

(Tag as any).Avatar = StyledAvatar;
(Tag as any).Close = Close;

/** @component */
export default Tag as React.ForwardRefExoticComponent<
  ITagProps & React.RefAttributes<HTMLDivElement>
> & {
  Avatar: typeof StyledAvatar;
  Close: typeof Close;
};
