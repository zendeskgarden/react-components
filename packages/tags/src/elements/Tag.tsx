/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { forwardRef, HTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import { StyledTag } from '../styled';
import { Close } from './Close';
import { Avatar } from './Avatar';

export interface ITagProps extends HTMLAttributes<HTMLDivElement> {
  /** Adjusts font size and padding */
  size?: 'small' | 'medium' | 'large';
  /**
   * Sets the color of the tag. Refer to
   * [PALETTE](/components/palette#palette)
   * for available colors. Accepts any hex value.
   */
  hue?: string;
  /** Applies pill styling */
  isPill?: boolean;
  /** Applies styles to round the tag */
  isRound?: boolean;
  /** Applies regular (non-bold) font weight */
  isRegular?: boolean;
}

const TagComponent = forwardRef<HTMLDivElement, ITagProps>(({ size, hue, ...otherProps }, ref) => (
  <StyledTag ref={ref} size={size} hue={hue} {...otherProps} />
));

TagComponent.displayName = 'Tag';

TagComponent.propTypes = {
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  hue: PropTypes.string,
  isPill: PropTypes.bool,
  isRound: PropTypes.bool,
  isRegular: PropTypes.bool
};

TagComponent.defaultProps = {
  size: 'medium'
};

/**
 * @extends HTMLAttributes<HTMLDivElement>
 */
export const Tag = TagComponent as typeof TagComponent & {
  Avatar: typeof Avatar;
  Close: typeof Close;
};

Tag.Avatar = Avatar;
Tag.Close = Close;
