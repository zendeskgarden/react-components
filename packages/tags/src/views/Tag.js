/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import PropTypes from 'prop-types';
import styled from 'styled-components';
import classNames from 'classnames';
import TagStyles from '@zendeskgarden/css-tags';

import { retrieveTheme, isRtl } from '@zendeskgarden/react-theming';

const COMPONENT_ID = 'tags.tag_view';

const TYPE = {
  GREY: 'grey',
  BLUE: 'blue',
  KALE: 'kale',
  RED: 'red',
  GREEN: 'green',
  YELLOW: 'yellow'
};

const SIZE = {
  SMALL: 'small',
  LARGE: 'large'
};

/**
 * Accepts all `<div>` props
 */
const Tag = styled.div.attrs({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  className: props =>
    classNames(TagStyles['c-tag'], {
      // Sizes
      [TagStyles['c-tag--sm']]: props.size === SIZE.SMALL,
      [TagStyles['c-tag--lg']]: props.size === SIZE.LARGE,

      // Shapes
      [TagStyles['c-tag--pill']]: props.pill,

      // Interaction States
      [TagStyles['is-focused']]: props.focused,
      [TagStyles['is-hovered']]: props.hovered,

      // Colors
      [TagStyles['c-tag--grey']]: props.type === TYPE.GREY,
      [TagStyles['c-tag--blue']]: props.type === TYPE.BLUE,
      [TagStyles['c-tag--kale']]: props.type === TYPE.KALE,
      [TagStyles['c-tag--red']]: props.type === TYPE.RED,
      [TagStyles['c-tag--green']]: props.type === TYPE.GREEN,
      [TagStyles['c-tag--yellow']]: props.type === TYPE.YELLOW,

      // RTL
      [TagStyles['is-rtl']]: isRtl(props)
    })
})`
  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

Tag.propTypes = {
  size: PropTypes.oneOf([SIZE.SMALL, SIZE.LARGE]),
  pill: PropTypes.bool,
  focused: PropTypes.bool,
  hovered: PropTypes.bool,
  type: PropTypes.oneOf([TYPE.GREY, TYPE.BLUE, TYPE.KALE, TYPE.RED, TYPE.GREEN, TYPE.YELLOW])
};

/** @component */
export default Tag;
