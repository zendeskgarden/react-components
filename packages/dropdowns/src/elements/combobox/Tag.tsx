/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLAttributes, forwardRef, useContext } from 'react';
import PropTypes from 'prop-types';
import { ThemeContext } from 'styled-components';
import { DEFAULT_THEME, useDocument, useText } from '@zendeskgarden/react-theming';
import { Tooltip } from '@zendeskgarden/react-tooltips';
import { ITagProps } from '../../types';
import useComboboxContext from '../../context/useComboboxContext';
import { StyledTag } from '../../views';
import { toString } from './utils';
import { TagAvatar } from './TagAvatar';

const TagComponent = forwardRef<HTMLDivElement, ITagProps>(
  ({ children, option, removeLabel, tooltipZIndex, ...props }, ref) => {
    const { getTagProps, isCompact, removeSelection } = useComboboxContext();
    const text = option.label || toString(option);
    const ariaLabel = useText(
      /* eslint-disable-next-line @typescript-eslint/no-use-before-define */
      Tag,
      props,
      'aria-label',
      `${text}, press delete or backspace to remove`,
      !option.disabled
    );
    const tagProps = getTagProps<HTMLDivElement>({
      option,
      'aria-label': ariaLabel!
    }) as HTMLAttributes<HTMLDivElement>;
    const theme = useContext(ThemeContext) || DEFAULT_THEME;
    const doc = useDocument(theme);
    const handleClick = () => removeSelection(option.value);

    return (
      <StyledTag
        aria-disabled={option.disabled}
        tabIndex={option.disabled ? undefined : 0}
        {...tagProps}
        {...props}
        size={isCompact ? 'medium' : 'large'}
        ref={ref}
      >
        {children || <span>{text}</span>}
        {!option.disabled &&
          (removeLabel ? (
            /* istanbul ignore next */
            <Tooltip appendToNode={doc?.body} content={removeLabel} zIndex={tooltipZIndex}>
              <StyledTag.Close aria-label={removeLabel} onClick={handleClick} />
            </Tooltip>
          ) : (
            <StyledTag.Close onClick={handleClick} />
          ))}
      </StyledTag>
    );
  }
);

TagComponent.displayName = 'Tag';

TagComponent.propTypes = {
  hue: PropTypes.string,
  isPill: PropTypes.bool,
  isRegular: PropTypes.bool,
  removeLabel: PropTypes.string
};

/**
 * @extends HTMLAttributes<HTMLDivElement>
 */
export const Tag = TagComponent as typeof TagComponent & {
  Avatar: typeof TagAvatar;
};

Tag.Avatar = TagAvatar;
