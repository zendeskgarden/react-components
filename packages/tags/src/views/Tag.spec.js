/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { render, renderRtl } from 'garden-test-utils';

import Tag from './Tag';

describe('Tag', () => {
  it('renders default styling', () => {
    const { container } = render(<Tag />);

    expect(container.firstChild).toHaveClass('c-tag');
  });

  it('renders RTL styling if provided', () => {
    const { container } = renderRtl(<Tag />);

    expect(container.firstChild).toHaveClass('is-rtl');
  });

  it('renders pill styling if provided', () => {
    const { container } = render(<Tag pill />);

    expect(container.firstChild).toHaveClass('c-tag--pill');
  });

  it('renders round styling if provided', () => {
    const { container } = render(<Tag round />);

    expect(container.firstChild).toHaveClass('c-tag--round');
  });

  describe('size', () => {
    it('renders small styling if provided', () => {
      const { container } = render(<Tag size="small" />);

      expect(container.firstChild).toHaveClass('c-tag--sm');
    });

    it('renders large styling if provided', () => {
      const { container } = render(<Tag size="large" />);

      expect(container.firstChild).toHaveClass('c-tag--lg');
    });
  });

  describe('state', () => {
    it('renders focused styling if provided', () => {
      const { container } = render(<Tag focused />);

      expect(container.firstChild).toHaveClass('is-focused');
    });

    it('renders hovered styling if provided', () => {
      const { container } = render(<Tag hovered />);

      expect(container.firstChild).toHaveClass('is-hovered');
    });
  });

  describe('visual types', () => {
    it('renders grey styling if provided', () => {
      const { container } = render(<Tag type="grey" />);

      expect(container.firstChild).toHaveClass('c-tag--grey');
    });

    it('renders blue styling if provided', () => {
      const { container } = render(<Tag type="blue" />);

      expect(container.firstChild).toHaveClass('c-tag--blue');
    });

    it('renders kale styling if provided', () => {
      const { container } = render(<Tag type="kale" />);

      expect(container.firstChild).toHaveClass('c-tag--kale');
    });

    it('renders red styling if provided', () => {
      const { container } = render(<Tag type="red" />);

      expect(container.firstChild).toHaveClass('c-tag--red');
    });

    it('renders green styling if provided', () => {
      const { container } = render(<Tag type="green" />);

      expect(container.firstChild).toHaveClass('c-tag--green');
    });

    it('renders yellow styling if provided', () => {
      const { container } = render(<Tag type="yellow" />);

      expect(container.firstChild).toHaveClass('c-tag--yellow');
    });

    it('renders fuschia styling if provided', () => {
      const { container } = render(<Tag type="fuschia" />);

      expect(container.firstChild).toHaveClass('c-tag--fuschia');
    });

    it('renders pink styling if provided', () => {
      const { container } = render(<Tag type="pink" />);

      expect(container.firstChild).toHaveClass('c-tag--pink');
    });

    it('renders crimson styling if provided', () => {
      const { container } = render(<Tag type="crimson" />);

      expect(container.firstChild).toHaveClass('c-tag--crimson');
    });

    it('renders orange styling if provided', () => {
      const { container } = render(<Tag type="orange" />);

      expect(container.firstChild).toHaveClass('c-tag--orange');
    });

    it('renders lemon styling if provided', () => {
      const { container } = render(<Tag type="lemon" />);

      expect(container.firstChild).toHaveClass('c-tag--lemon');
    });

    it('renders lime styling if provided', () => {
      const { container } = render(<Tag type="lime" />);

      expect(container.firstChild).toHaveClass('c-tag--lime');
    });

    it('renders mint styling if provided', () => {
      const { container } = render(<Tag type="mint" />);

      expect(container.firstChild).toHaveClass('c-tag--mint');
    });

    it('renders teal styling if provided', () => {
      const { container } = render(<Tag type="teal" />);

      expect(container.firstChild).toHaveClass('c-tag--teal');
    });

    it('renders azure styling if provided', () => {
      const { container } = render(<Tag type="azure" />);

      expect(container.firstChild).toHaveClass('c-tag--azure');
    });

    it('renders royal styling if provided', () => {
      const { container } = render(<Tag type="royal" />);

      expect(container.firstChild).toHaveClass('c-tag--royal');
    });

    it('renders purple styling if provided', () => {
      const { container } = render(<Tag type="purple" />);

      expect(container.firstChild).toHaveClass('c-tag--purple');
    });
  });
});
