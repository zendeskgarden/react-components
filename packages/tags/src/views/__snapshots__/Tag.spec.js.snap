// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag renders RTL styling if provided 1`] = `
<div
  className="c-tag is-rtl "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
/>
`;

exports[`Tag renders default styling 1`] = `
<div
  className="c-tag "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
/>
`;

exports[`Tag renders pill styling if provided 1`] = `
<div
  className="c-tag c-tag--pill "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
/>
`;

exports[`Tag size renders large styling if provided 1`] = `
<div
  className="c-tag c-tag--lg "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
  size="large"
/>
`;

exports[`Tag size renders small styling if provided 1`] = `
<div
  className="c-tag c-tag--sm "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
  size="small"
/>
`;

exports[`Tag state renders focused styling if provided 1`] = `
<div
  className="c-tag is-focused "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
/>
`;

exports[`Tag state renders hovered styling if provided 1`] = `
<div
  className="c-tag is-hovered "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
/>
`;

exports[`Tag visual types renders blue styling if provided 1`] = `
<div
  className="c-tag c-tag--blue "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
  type="blue"
/>
`;

exports[`Tag visual types renders green styling if provided 1`] = `
<div
  className="c-tag c-tag--green "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
  type="green"
/>
`;

exports[`Tag visual types renders grey styling if provided 1`] = `
<div
  className="c-tag c-tag--grey "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
  type="grey"
/>
`;

exports[`Tag visual types renders kale styling if provided 1`] = `
<div
  className="c-tag c-tag--kale "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
  type="kale"
/>
`;

exports[`Tag visual types renders red styling if provided 1`] = `
<div
  className="c-tag c-tag--red "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
  type="red"
/>
`;

exports[`Tag visual types renders yellow styling if provided 1`] = `
<div
  className="c-tag c-tag--yellow "
  data-garden-id="tags.tag_view"
  data-garden-version="version"
  type="yellow"
/>
`;
